AED-766 Gasless NFT minting (#27)
OpenZeppelin/defender-templates: defender/gasless-nft-minting/README.md

# Gasless NFT Minter

This code will allow users to deploy an [ERC-2771-compatible](https://docs.openzeppelin.com/contracts/4.x/api/metatx#ERC2771Context) [ERC-1155 NFT](https://docs.openzeppelin.com/contracts/4.x/erc1155) contract and then use one Autotask to sign mint requests, and another Autotask to relay the request to a [trusted forwarder](https://docs.openzeppelin.com/contracts/4.x/api/metatx#MinimalForwarder) which will send the request to the NFT contract to mint an NFT for the user.

## Setup

This section will help you set up the following:
- Acquire Defender and Relay keys
- Add funds to the Relay
- Deploy contracts
  - `yarn deploy contract --contract-name MinimalForwarder`
  - `yarn deploy contract --contract-name DemoNFT "0xMINIMAL_FORWARDER_ADDRESS"`
- Update `config.dev.yml` with the 2 deployed contract addresses
- Deploy the Autotasks with `serverless deploy`
- Connect Relays to deployed Autotasks

>   - Contract Deployer relay must be connected to the signing Autotask

Reviewer: ```suggestion
  - Contract Deployer Relay must be connected to the signing Autotask
```

  - Any Relayer with funds can be connected to the Relay Autotask
- Copy the 2 Autotask webhooks into the `config.dev.yml` file

### Defender Account Setup

- In your [Defender account](https://defender.openzeppelin.com/), select the Hamburger icon in the upper right corner and click on **Team API Keys**
- In the Team API Keys screen, click **Create API Key**

> - Make sure that the options for **Manage Autotasks** and **Manage Proposals and Contracts** are selected (we do not make use of the **Manage Relayers** and **Manage Sentinels** options with this template)

Reviewer: When I only selected "Manage Autotasks" and "Manage Proposals and Contracts" I could not successfully deploy with Serverless. Here is the error output I received:
```shell
❯ serverless deploy --stage dev
Running "serverless" from node_modules
(node:66347) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.

Please migrate your code to use AWS SDK for JavaScript (v3).
For more information, check the migration guide at https://a.co/7PzMCcy
(Use `node --trace-warnings ...` to show where the warning was created)
========================================================
Secrets
✔ Updated Secret (gasless_nft_minting_part_deux_FORWARDER_ADDRESS)
✔ Updated Secret (gasless_nft_minting_part_deux_NFT_ADDRESS)
Contracts
Relayers
Autotasks
✖ User is not authorized to access this resource
✖ User is not authorized to access this resource
Notifications
Environment: darwin, node 18.12.1, framework 3.27.0 (local) 3.27.0v (global), plugin 6.2.3, SDK 4.3.2
Docs:        docs.serverless.com
Support:     forum.serverless.com
Bugs:        github.com/serverless/serverless/issues

Error:
DefenderApiResponseError: Request failed with status code 403
    at rejectWithDefenderApiError (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-base-client/lib/api/api.js:11:27)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-sentinel-client/lib/api/index.js:85:20
    at async SentinelClient.apiCall (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-base-client/lib/api/client.js:26:20)
    at async DefenderDeploy.deployCategories (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-serverless/lib/cmd/deploy.js:381:31)
    at async DefenderDeploy.deploy (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-serverless/lib/cmd/deploy.js:775:9)
    at async DefenderDeploy.requestConfirmation (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/defender-serverless/lib/cmd/deploy.js:131:9)
    at async PluginManager.runHooks (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/lib/classes/plugin-manager.js:530:9)
    at async PluginManager.invoke (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/lib/classes/plugin-manager.js:564:9)
    at async PluginManager.run (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/lib/classes/plugin-manager.js:604:7)
    at async Serverless.run (/Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/lib/serverless.js:179:5)
    at async /Users/echo/Documents/openzeppelin/eth-denver-2023-templates/defender/gasless-nft-minting/node_modules/serverless/scripts/serverless.js:812:9
```

When I created an API key that allowed me to manage all of the Defender resources, the deployment worked.

Author: Changed it to add all permissions 9daa9144369044130f497f8c4a41219aab381ec9

  - In the Relay Dashboard screen, click **Create Relayer** in the upper right corner
  - Pick a meaningful name, choose the network that you want to deploy to, and click **Create**
- Click on the existing Relay that you would like to use for deployments
- Click on the settings gear/cog **⚙** at the top middle and choose **+ Create new API key**
- Copy your Ethereum Address, API key, and Secret key to a local file (you will **NOT** be able to view your Secret key again after this message box goes away)
- Make sure that you really did copy your API key and Secret key to a local file
- Check the box for **I’ve written down the secret key** and select **Close**

### Add Funds to the Relay

- In your [Defender account](https://defender.openzeppelin.com/)
- Click on the **Relay** link on the left side
- Find the Relay that deployed the NFT contract, click on the **copy address** button
- Transfer funds to that address to cover the deployment and NFT minting costs

### Local Code Setup

In the `gasless-nft-minting` directory, perform the following steps:

- A secrets file can be created for each stage of production. We will create one for development:

>   - Copy and rename the `sample.secrets.yml` to `.secrets/dev.yml`

Reviewer: The `sample.secrets.yml` file is not in the `gasless-nft-minting` directory. It is in the `defender` directory.

Author: Fixed

Author: 9daa9144369044130f497f8c4a41219aab381ec9

  - Modify the four lines in the `.secrets/dev.yml` file, replacing the portion in the angle brackets `<>` with your Defender API keys and Secret keys, as indicated:

>   - `defender-api-key: <API Key goes here>`
>   - `defender-api-secret: <Secret key goes here>`
>   - `relay-api-key: <Relay API Key goes here>`
>   - `relay-api-secret: <Relay Secret key goes here>`

Reviewer: ```suggestion
    - `defender-api-key: <API Key goes here>`
    - `defender-api-secret: <Secret key goes here>`
    - `relay-api-key: <Relay API Key goes here>`
    - `relay-api-secret: <Relay Secret key goes here>`
```

### Deploying Contracts

Contracts will be deployed using the Defender Relay that you configured earlier.

- The `MinimalForwarder` contract must be deployed first
  - `yarn deploy contract --contract-name MinimalForwarder`
- Take the address of the `MinimalForwarder` and provide it as a constructor for the NFT contract
  - `yarn deploy contract --contract-name DemoNFT "0xMINIMAL_FORWARDER_ADDRESS"`
- Make a note of both newly deployed contract addresses. They will need to be added to the config file mentioned below

### Serverless Deployment Setup :rocket:

Now that the contracts are deployed, they can be added to the config file so that the Autotasks know which contracts they are signing for and forwarding to. In the `gasless-nft-minting` directory, perform the following steps:

- A config file can be created for each stage of production. We will create one for development

>   - Copy and rename `sample.config.yml` to `config.dev.yml`

Reviewer: ```suggestion
  - Copy and rename `example.config.yml` to `config.dev.yml`
```

Author: renamed the file. 9daa9144369044130f497f8c4a41219aab381ec9

  - Open the file and modify the following values with the :
    - `forwarder-address`
    - `nft-address`
- Run `serverless deploy --stage dev` to deploy the stack to Defender

### Connecting Relays to Deployed Autotasks

Since the contracts were deployed with Defender, we can simply use the same Relayer (current owner of the contract) to handle the signing of the NFT mint requests. The Relayer must be reconnected after each deployment.

- In your [Defender account](https://defender.openzeppelin.com/)
- Click on the **Autotask** link on the left side
- Click on **Gasless NFT Minting Signer Autotask** Autotask
- Click on the settings **gear/cog** at the top middle, and select **Edit Settings**
- In the **Connect a Relayer** section, choose the Relayer that was used to deploy the contract

The same thing must be done for the Autotask that will relay the request:

- Click on the **Autotask** link on the left side
- Click on **Gasless NFT Minting Relayer Autotask** Autotask
- Click on the settings **gear/cog** at the top middle, and select **Edit Settings**

> - In the **Connect a Relayer** section, choose ANY Relayer that has enough funds to execute the minting requests on that are submitted to it

Reviewer: ```suggestion
- In the **Connect a Relayer** section, choose ANY Relayer that has enough funds to execute the minting requests that are submitted to it
```

### Get the Autotask webhook URLs

Interactions with the Autotasks will be done via webhook URLs. Add the URLs to your config file:

- In your [Defender account](https://defender.openzeppelin.com/)
- Click on the **Autotask** link on the left side
- Save the Signer webhook
  - In the **Gasless NFT Minting Signer Autotask** Autotask box, click on the **Copy** button to copy the Autotask Webhook URL to your clipboard
  - Save the Autotask URL as the value `signer-webhook` in your `config.dev.yml` file
- Save the Relayer webhook
  - In the **Gasless NFT Minting Relayer Autotask** Autotask box, click on the **Copy** button to copy the Autotask Webhook URL to your clipboard
  - Save the Autotask URL as the value `relayer-webhook` in your `config.dev.yml` file

## Typical Usage

The Signer Autotask can now authorize minting of Tokens without spending any gas by signing a minting request and sending it back to the requestor via HTTP.

A user with a signed request can send the request to the Relayer Autotask via HTTP and also not spend gas.

> The Relayer Autotask verify the request and signature and execute the transaction on-chain if they are valid.

Reviewer: ```suggestion
The Relayer Autotask verifies the request and signature and executes the transaction on-chain if they are valid.
```

### Signing and Relaying with scripts

> A `sign` and `relay` script are included in this template use the deployed Autotasks. 

Reviewer: ```suggestion
The `sign` and `relay` scripts included in this template use the deployed Autotasks. 
```

In the `gasless-nft-minting` directory, perform the following steps:
- `yarn sign <address>` - to have the Signer Autotask sign a mint request
- `yarn relay '<stringified JSON>'` - to relay the request and mint the NFT to the user

### Submitting Requests with Curl

> Signature requests can be submitted with curl. Replace the URL with your Signer Autotask webhook URL and 

Reviewer: ```suggestion
Signature requests can be submitted with curl. Replace the URL with your Signer Autotask webhook URL and the address with the recipient address:
```

```shell
curl -X POST https://api.defender.openzeppelin.com/autotasks/xxxx/runs/webhook/xxxx \
   -H 'Content-Type: application/json' \
   -d '{"address": "0xRECIPIENT_ADDRESS"}'
```

> Relay requests submitted with curl. Replace the URL with your Relay Autotask webhook URL

Reviewer: ```suggestion
Relay requests can also be submitted with curl. Replace the URL with your Relay Autotask webhook URL
```

```shell
curl -X POST https://api.defender.openzeppelin.com/autotasks/xxxx/runs/webhook/xxxx \
   -H 'Content-Type: application/json' \
   -d '{
    "signature":"0xVERY_LONG_SIGNATURE_STRING",
    "request":{
      "value":0,
      "gas":1000000,
      "nonce":"0",
      "to":"0xNFT_ADDRESS",
      "from":"0xRELAY_ADDRESS","data":"0x731133e9000000000000000000000000USER_ADDRESS00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000"
    }
  }'
```

## Tests

> Check that contract can mint with both standard and gasless transactions

Reviewer: ```suggestion
Check that the contract can mint with both standard and gasless transactions
```

- `yarn test tests/nftContract.test.js`
Check that the configured Defender Relay can sign and verify meta transactions
- `yarn test tests/eip712sig.online.test.js`

## Notes

> This is a proof of concept template, the Autotask currently allow for unlimited mints of NFTs and can be easily abused until the Relay runs out of funds.

Reviewer: ```suggestion
This is a proof of concept template. The Autotask currently allows for unlimited mints of NFTs and can be easily abused until the Relay runs out of funds.
```

### Relay API Keys

- In your [Defender account](https://defender.openzeppelin.com/), select the **◎ Relay** link on the upper left side to get to the Relay Dashboard
- If you do not have a Relay on the network that you want to deploy to, create one:
  - In the Relay Dashboard screen, click **Create Relayer** in the upper right corner
  - Pick a meaningful name, choose the network that you want to deploy to, and click **Create**
- Click on the existing Relay that you would like to use for deployments
- Click on the settings gear/cog **⚙** at the top middle and choose **+ Create new API key**
- Copy your Ethereum Address, API key, and Secret key to a local file (you will **NOT** be able to view your Secret key again after this message box goes away)
- Make sure that you really did copy your API key and Secret key to a local file
- Check the box for **I’ve written down the secret key** and select **Close**

### Add Funds to the Relay

- In your [Defender account](https://defender.openzeppelin.com/)
- Click on the **Relay** link on the left side
- Find the Relay that deployed the NFT contract, click on the **copy address** button

> - Transfer funds to that address to cover the deployment and NFT minting costs

Reviewer 2: I think this could be a bit misleading. At this point, it's saying to find the Relay that has already deployed the NFT contract, but then saying to transfer funds to the Relay address to cover the cost of deployment.

You could replace line 46 with something like `Find the Relay that will be used to deploy the NFT contract`

Author: Thanks, it has been updated.
